First pass at Quarkus 3 announcement (#1686)
quarkusio/quarkusio.github.io: _posts/2023-04-26-quarkus-3-0-final-released.adoc

---
layout: post
title: 'Quarkus 3.0 released'
date: 2023-04-26
tags: release
synopsis: 'It is our great pleasure to announce our new major release of Quarkus: Quarkus 3.0.'
author: gsmet
---

It is our great pleasure to announce the release of Quarkus 3.0,

> our new major release,

Reviewer: ```suggestion
our new major release,
```

our largest release ever,

> our new major release,

Reviewer: ```suggestion
our largest release ever,
```

which comes packed with new features and enhancements.

> The development of Quarkus 3 started on March 18 last year with the crazy idea

Reviewer: ```suggestion
The development of Quarkus 3 started on March 18 2022 with the crazy idea
```

that we could rewrite the whole tree to migrate to the new `jakarta.*` packages
while continuing the development of Quarkus 2.

Quarkus 3 is now based on Jakarta EE 10 and, as you can expect,
it is not the only change, this version comes packed with new features and enhancements
that we implemented during this period.

This announcement contains a lot of information about the new features and how to upgrade
so we recommend to read it fully before upgrading to Quarkus 3.

== What's new?

> === Jakarta EE 10

Reviewer 2: I wouldn't be so concerned about it being on top of list - its after all the main reason why we went with a 3.x rather than 2.x.

Its also better to compare to Spring 6 announcement which is what actually drove the jakarta change. here its the first thing. https://spring.io/blog/2022/11/16/spring-framework-6-0-goes-ga

> === Jakarta EE 10

Reviewer: Like I said, I understand where this is coming from, I would just do it differently myself :)

Quarkus 2 was based on Jakarta EE 8, Quarkus 3 is based on Jakarta EE 10.

The most visible change is that the `javax.*` packages from Jakarta EE has been moved to `jakarta.*`.
But Jakarta EE 10 also comes with new features for your favorite APIs.

[NOTE]
====
Only the `javax.*` packages from Jakarta EE have changed.
The ones from the JDK are still under `javax.*`.
====

=== Eclipse MicroProfile 6

TODO

=== Hibernate ORM 6

> This is one of the biggest changes in Quarkus 3: we upgraded to Hibernate ORM 6, more precisely 6.2.

Reviewer: ```suggestion
This is one of the biggest changes in Quarkus 3: we upgraded to Hibernate ORM 6, version 6.2 to be more precise.
```

Hibernate ORM 6 is a major new version of Hibernate ORM and it comes with a lot of changes,

> with some being breaking changes.

Reviewer: ```suggestion
some of them being breaking ones.
```

To learn more about the changes in Hibernate ORM 6, please have a look at the announcements for:

- https://in.relation.to/2022/03/31/orm-60-final/[Hibernate ORM 6.0]
- https://in.relation.to/2022/06/14/orm-61-final/[Hibernate ORM 6.1]

> - https://in.relation.to/2023/03/30/orm-62-final/[Hibernate ORM 6.2]

Reviewer: Do we really want to point users to three 3 different links? I thought we had a Wiki page, no?

Reviewer 2: agree - keep it simple first, details can be found for those who need to go deep.

Upgrading to Hibernate ORM 6 will require some efforts and testing, make sure you have a look at the <<upgrading>> section below.

> Some improvements have also been made to the extension itself such as:

Reviewer: I think we should also add the `StatelessSession` injection here.

Reviewer: Ah sorry, missed it 🤦🏼

- The support of `StatementInspector` as `@PersistenceUnitExtension` managed bean
- The ability to inject a https://docs.jboss.org/hibernate/orm/6.2/userguide/html_single/Hibernate_User_Guide.html#_statelesssession[`StatelessSession`] via CDI

=== Hibernate Reactive 2

Hibernate Reactive has been upgraded to version 2 to be compatible with Hibernate ORM 6.

It is still in Beta for now but should be promoted to Final in the coming months.

=== New Dev UI

> Quarkus 2 introduced the Dev UI which provides a web ui to use during development.

Reviewer: ```suggestion
Quarkus 2 introduced the Dev UI which provides a web UI to use during development.
```

Extensions could provide their own Dev UI pages to provide additional functionality.

Quarkus 3 now has a new Dev UI which is more extensible and easier to use.

> It also has a new look and feel.

Reviewer: ```suggestion
It also has a new - much improved look and feel.
```

The new Dev UI is the default but,
given all the extensions have not yet migrated to the new one,
you can still access the old one at `/q/dev-v1` but it will be dropped in a future version.

We have a https://www.youtube.com/watch?v=sz5ihmA4gaE&list=PLsM3ZE5tGAVbyncLm7ue2V25cwFck7ew9[playlist] on Quarkus YouTube channel that shows off the new Dev UI, explains the new features and how to use and extend it.

=== CLI plugins

The Quarkus CLI was made extensible with plugins.
Expect a lot of new features in the Quarkus CLI in the form of plugins in the near future.

=== Upgrade tool

Quarkus 3 comes with an upgrade tool that will radically simplify the upgrades from one version to another.

You can learn more about it in the <<upgrading>> section.

=== Support for Maven 3.9

> Maven 3.9 came with some breaking changes and the integration with Quarkus was not working properly.

Reviewer: ```suggestion
Maven 3.9 brought some breaking changes and as a result, the integration with Quarkus was not working properly.
```

> It is now fixed and you can use Maven 3.9 to build your Quarkus 3.0.0.CR1 applications.

Reviewer 3: CR1 -> Final

=== Gradle 8

Our Gradle support has been upgraded to Gradle 8.

=== Management network interface

> Until now, all the Quarkus endpoints were exposed on the same network interface.

Reviewer: ```suggestion
Hitherto, all the Quarkus endpoints were exposed on the same network interface.
```

Reviewer 2: didnt apply this as hitherto isn't that commonly used I would say.

Reviewer: We could even do `Heretofore` if you need more exocit 😆

All without requiring changes to the project dependencies or configuration and therefore simplifying developer experience.

=== /q/info

To expose information about your application (such as the git hash), add the `quarkus-info` extension to your project.

The endpoint is available on `/q/info` and will be exposed on the management network interface if you enable it.

=== RESTEasy Reactive

A lot of usability enhancements have come into RESTEasy Reactive such as the ability to retrieve all the Multipart parts.

Remember that RESTEasy Reactive is our REST layer covering both reactive and blocking workloads.

=== OpenTelemetry

The OpenTelemetry extension has been rewritten to support the SDK autoconfiguration and went under a lot of changes.

The configuration namespace has changed to `quarkus.otel.*` and it is recommended to switch to the new configuration properties, even if the old ones are still supported for now.

Enabling OpenTelemetry for JDBC is now as simple as setting `quarkus.datasource.jdbc.telemetry` to `true`.
You don't have to modify your JDBC connection URL anymore.

=== Multiple mailers

> Sending emails via several SMTP servers is supported in Quarkus 3.0.0.CR1.

Reviewer 3: CR1 -> Final

Have a look at the https://quarkus.io/version/main/guides/mailer-reference#multiple-mailer-configurations[updated documentation].

=== Qute

Qute, our templating engine, also got a lot of love with several new features such as the ability to cache a section of the template that rarely changes thanks to link:/guides/qute-reference#cached-section[cached sections].

=== Cache

It is now possible to use a Redis backend with the Cache extension.

More information in the https://quarkus.io/version/main/guides/cache-redis-reference[dedicated guide].

The cache extension now allows you to define global defaults cache configuration that will be applied to all your caches.

=== Database migrations

> Your database migrations with Flyway and Liquibase are now run as init containers in manifests.

Reviewer: @iocanel do we have a link to some our docs for this feature?

The Flyway extension supports custom credentials/URL to connect to the database
and you can more easily customize the configuration of the Flyway instance.

=== MongoDB

> ``CredentialsProvider``s are now supported for MongoDB connections.

Reviewer: ```suggestion
`CredentialsProvider`s are now supported for MongoDB connections.
```

so that you have the time to upgrade your applications to Quarkus 3.x.

== Full changelog

You can get the full changelog of Quarkus 3.0 on GitHub:

* 3.0.0.Alpha1 and 3.0.0.Alpha2 changelogs are empty as these versions were just a transformation of Quarkus 2 versions to Jakarta EE 10
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha3[3.0.0.Alpha3]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha4[3.0.0.Alpha4]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha5[3.0.0.Alpha5]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha6[3.0.0.Alpha6]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Beta1[3.0.0.Beta1]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.CR1[3.0.0.CR1]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.CR2[3.0.0.CR2]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Final[3.0.0.Final]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.1.Final[3.0.1.Final]

== Contributors

The Quarkus community is growing and has now https://github.com/quarkusio/quarkus/graphs/contributors[788 contributors].
Many many thanks to each and everyone of them.

> In particular for the 3.0 release, thanks to Adler Fleurant, Adrian Pauli, Ales Justin, Alex Martel, Alexandre Dutra, Alexei Bratuhin, Alexey Loubyansky, Alexey Sharandin, amoscatelli, Andrea Cosentino, Andrea Peruffo, Andri Reveli, Andy Damevin, Àngel Ollé Blázquez, Antonio Costa, Antonio Goncalves, Antonio Jacob Costa, arik-dig, Ashish Ranjan, Auri Munoz, Benedikt Ritter, benstone, besta, Bill Burke, Brad Hards, Bruno Baptista, Bruno Borges, Bruno Leonardo Gonçalves, Bård Kristian Haaland-Sørensen, Carles Arnal, Chexpir, Chihiro Ito, Chris Laprun, Christian Berger, Christian Pieczewski, Christian von Atzigen, Christoph Hermann, Clemens Classen, Clement Escoffier, Damon Sutherland, Dan Dunning, David Andlinger, David Arnold, David M. Lloyd, Davide D'Alto, Dmitri Bourlatchkov, Efim Smykov, Eric Deandrea, Erin Schnabel, faculbsz, Falko Modler, Felipe Henrique Gross Windmoller, Fikru  Mengesha, Filippe Spolti, Foivos Zakkak, Foobartender, Fouad Almalki, franz1981, Galder Zamarreño, Geert Schuring, George Gastaldi, Georgios Andrianakis, Gerhard Flothow, Giovanni Barbaro, Guillaume DOUSSIN, Guillaume Le Floch, Guillaume Nieser, Guillaume Smet, Gunnar Morling, Gwenneg Lepage, Harald Albers, Helber Belmiro, Holly Cummins, Hugo Guerrero, IgnasiCR, imperatorx, Ioannis Canellos, Ivan Bazalii, Jan Martiska, Jan Wiemer, Jasmin Suljic, Jayson Minard, Jesse Ehrenzweig, Joe Siponen, John OHara, Jonathan Kolberg, Jorge Solórzano, Jose, Jose Carvajal, Josef Andersson, jtama, Julien Ponge, Julio Enrique Santana Lora, Justin Lee, Katia Aresti, kdnakt, Kevin Dubois, Konstantin Gribov, Konstantin Silin, kottmann, Ladislav Thon, Laure Souche, Leandro Quiroga, Loïc Mathieu, luca-bassoricci, Lukas Lowinger, Lukáš Petrovický, luneo7, Manyanda Chitimbo, Marc Nuri, Marc Schlegel, Marcel Lohmann, Marcell Cruz, Marco Bungart, Marco Schaub, Marek Skacelik, marko-bekhta, Markus Himmel, Martin Kouba, Martin Panzer, Marvin B. Lillehaug, Matej Novotny, Matteo Mortari, Max Rydahl Andersen, Mazen Khalil, Melloware, mfpc, Michael Edgar, Michael Mosmann, Michael Musgrove, Michal Karm Babacek, Michal Maléř, Michal Vavřík, Michelle Purcell, Mickey Maler, Mihai.Poenaru, Moritz Heine, mrizzi, mun711, Nathan Erwin, Nathan Mittelette, Nicolas Filotto, nscuro, oliv37, Orbifoldt, Ozan Gunalp, Özkan Pakdil, Pablo Gonzalez Granados, Paulo Casaes, Pavel.Vervenko, Pedro Igor, Pedro Pereira, Peter Palaga, Phillip Krüger, Radoslaw Adamiak, Radovan Synek, Ramon Boss, Robbie Gemmell, Robert Gonciarz, Robert Stupp, Roberto Cortez, Rolfe Dlugy-Hegwer, Romain Quinio, Rostislav Svoboda, Sanne Grinovero, Sebastian Schuster, Sergey Beryozkin, Severin Gehwolf, shjones, skraft-redhat, spencercjh, Stuart Douglas, sturdy5, Stéphane Épardaud, Sébastien CROCQUESEL, Sébastien Crocquesel, Tamas Cservenak, Theodor Mihalache, Thomas Segismont, Thor Weinreich, tom, Tom Cunningham, Vaclav Svejcar, Vincent Sevel xstefank, Yoann Rodière, Yoshikazu Nojima, Yubao Liu, zedbeit, Zheng Feng, and Žiga Deisinger.

Reviewer 4: Would it be possible to add a link to their GitHub accounts here?

---
layout: post
title: 'Quarkus 3.0, our new major release, is here!'
date: 2023-04-26
tags: release
synopsis: 'It is our great pleasure to announce our new major release of Quarkus: Quarkus 3.0.'
author: gsmet
---

> Quarkus 3.0 is the result of a lot of work and dedication from the community. Quarkus continues to be an Open Source stack to write Java applications offering unparalleled startup time, memory footprint and developer experience.

Reviewer 5: ```suggestion
Quarkus 3.0 is the result of a lot of work and dedication from the community. Quarkus continues to be an Open Source stack to write Java applications offering unparalleled startup time, memory footprint, and developer experience.
```

> The development of Quarkus 3 started last year on March 18 2022 with the crazy idea

Reviewer 5: ```suggestion
The development of Quarkus 3 started last year on March 18, 2022, with the crazy idea
```

that we could rewrite the whole tree to migrate to the new `jakarta.*` packages
while continuing the development of Quarkus 2.

Quarkus 3.0 is based on Jakarta EE 10 and, as you can expect,

> it is not the only change, this version comes packed with new features and enhancements

Reviewer 5: ```suggestion
it is not the only change. This version comes packed with new features and enhancements
```

that we implemented during this period.

This announcement contains a lot of information about the new features and how to upgrade

> so we recommend to read it fully before upgrading to Quarkus 3.

Reviewer 5: ```suggestion
so we recommend reading it thoroughly before upgrading to Quarkus 3.
```

=== OpenTelemetry

The OpenTelemetry extension has been rewritten to support the SDK autoconfiguration and went under a lot of changes.

The configuration namespace has changed to `quarkus.otel.*` and it is recommended to switch to the new configuration properties, even if the old ones are still supported for now.

Also, enabling OpenTelemetry for JDBC is now as simple as setting `quarkus.datasource.jdbc.telemetry` to `true`.
You don't have to modify your JDBC connection URL anymore.

=== Multiple mailers

Sending emails via multiple SMTP servers is supported in Quarkus 3.0.

Have a look at the https://quarkus.io/version/main/guides/mailer-reference#multiple-mailer-configurations[updated documentation].

=== Qute

Qute, our templating engine, also got a lot of love with several new features such as the ability to cache a section of the template that rarely changes thanks to link:/guides/qute-reference#cached-section[cached sections].

=== Cache

It is now possible to use a Redis backend with the Cache extension.

More information in the https://quarkus.io/version/main/guides/cache-redis-reference[dedicated guide].

> The cache extension also allows you to define global defaults cache configuration that will be applied to all your caches.

Reviewer 5: ```suggestion
The cache extension also allows you to define a global default cache configuration that will be applied to all your caches.
```

=== Database migrations

Your database migrations with Flyway and Liquibase are now run as init containers in manifests.

> The notion of migration/setup work being done in init cotainers is available for other extensions to implement and support.

Reviewer 5: ```suggestion
The notion of migration/setup work being done in init containers is available for other extensions to implement and support.
```

> The Flyway extension supports custom credentials/URL to connect to the database

Reviewer 5: ```suggestion
The Flyway extension supports custom credentials/URLs to connect to the database
```

and you can more easily customize the configuration of the Flyway instance.

=== MongoDB

``CredentialsProvider``s are now supported for MongoDB connections.

=== Elasticsearch Java Client extension

> The new Elasticsearch Java Client is supported as a brand new extension.

Reviewer 5: ```suggestion
The new Elasticsearch Java Client is supported as a brand-new extension.
```

> This solves the licensing problems that prevented us to update the deprecated High Level REST Client.

Reviewer 5: ```suggestion
This solves the licensing problems that prevented us from updating the deprecated High Level REST Client.
```

To use this new client, have a look at the updated link:/guides/elasticsearch[Elasticsearch guide].

=== gRPC

> Several enhancements have been made to the gRPC extensions such as the support of `InProcess`.

Reviewer 5: ```suggestion
Several enhancements have been made to the gRPC extensions, such as the support of `InProcess`.
```

=== Scheduler API

You can now schedule jobs programmatically by using the Scheduler programmatic API,
described in the link:/guides/scheduler-reference#programmatic_scheduling[Scheduler reference guide].

=== Kubernetes Client

The Kubernetes Client has been upgraded to version 6.5.

=== Azure Functions extension

> The development of Azure functions is easier than ever thanks to the new Azure Functions extension.

Reviewer 5: ```suggestion
The development of Azure functions is easier than ever, thanks to the new Azure Functions extension.
```

Learn more about it in the link:/guides/azure-functions[dedicated guide]

== Other changes

=== Java 11 deprecated

> The OpenJDK community will end active support for Java 11 in September 2023. We still plan to support Java 11 past that date for core Quarkus functionality but Java 11 is now marked as deprecated. We recommend that you upgrade to Java 17 or later if you want to use the latest and greatest version of Quarkus.

Reviewer 5: ```suggestion
The OpenJDK community will end active support for Java 11 in September 2023. We still plan to support Java 11 past that date for core Quarkus functionality, but Java 11 is now marked as deprecated. We recommend that you upgrade to Java 17 or later if you want to use the latest and greatest version of Quarkus.
```

=== Release cadence and Long Term Support

> With Quarkus 3 finally out, we will be returning to our regular continuous cadence of releasing approximately every 5 weeks. This provides a delivery train of small incremental changes that are easy to adopt and upgrade to.

Reviewer 5: ```suggestion
With Quarkus 3 finally out, we will be returning to our regular, continuous cadence of releasing approximately every five weeks. This provides a delivery train of small incremental changes that are easy to adopt and upgrade to.
```

> We do know some of you are looking for a more stable release cadence and we are working on a new long term support (LTS) policy starting from Quarkus 3.2. We will provide details on this as we get closer to the 3.2 release. Java 11 will still be supported there for the core part of Quarkus.

Reviewer 5: ```suggestion
We do know some of you are looking for a more stable release cadence, and we are working on a new long-term support (LTS) policy starting from Quarkus 3.2. We will provide details on this as we get closer to the 3.2 release. Java 11 will still be supported there for the core part of Quarkus.
```

=== Quarkiverse

> Quarkus is not just the https://github.com/quarkusio/quarkus repository and the Quarkus platform it is also the whole rest of the Quarkus ecosystem - the so called Quarkiverse. Those extensions are hosted and maintained by lots of contributors across the world and across companies - some of them host them in their own GitHub repositories, some of them host them in the https://github.com/quarkiverse[Quarkiverse Hub].

Reviewer 5: ```suggestion
Quarkus is not just the https://github.com/quarkusio/quarkus repository and the Quarkus platform. It is also the whole rest of the Quarkus ecosystem - the so-called Quarkiverse. Those extensions are hosted and maintained by lots of contributors and organizations around the world. Some host these extensions in their GitHub repositories, and others host them in the https://github.com/quarkiverse[Quarkiverse Hub].
```

> With Quarkus 3 due mainly to the package changes in many of the core dependencies, we are happy to say that lots of those extensions have already been updated to work with Quarkus 3.0.

Reviewer 5: ```suggestion
With Quarkus 3, due mainly to the package changes in many of the core dependencies, we are happy to say that lots of those extensions have already been updated to work with Quarkus 3.0.
```

That's why we will maintain Quarkus 2.16 with bugfixes and important CVE fixes for a few months
so that you have the time to upgrade your applications to Quarkus 3.x.

== Full changelog

You can get the full changelog of Quarkus 3.0 on GitHub:

* 3.0.0.Alpha1 and 3.0.0.Alpha2 changelogs are empty as these versions were just a transformation of Quarkus 2 versions to Jakarta EE 10
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha3[3.0.0.Alpha3]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha4[3.0.0.Alpha4]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha5[3.0.0.Alpha5]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Alpha6[3.0.0.Alpha6]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Beta1[3.0.0.Beta1]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.CR1[3.0.0.CR1]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.CR2[3.0.0.CR2]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.0.Final[3.0.0.Final]
* https://github.com/quarkusio/quarkus/releases/tag/3.0.1.Final[3.0.1.Final]

== Contributors

> The Quarkus community is growing and has now https://github.com/quarkusio/quarkus/graphs/contributors[788 contributors].

Reviewer 5: ```suggestion
The Quarkus community is growing and now has https://github.com/quarkusio/quarkus/graphs/contributors[788 contributors].
```

> Many many thanks to each and everyone of them.

Reviewer 5: ```suggestion
Many, many thanks to each and every one of them.
```

We also want to thank all the extension maintainers from the Quarkiverse and beyond who worked hard on getting the extension ecosystem ready for Quarkus 3.0.

== Come Join Us

> We value your feedback a lot so please report bugs, ask for improvements... Let's build something great together!

Reviewer 5: ```suggestion
We value your feedback a lot, so please report bugs, and ask for improvements... Let's build something great together!
```

If you are a Quarkus user or just curious, don't be shy and join our welcoming community:

>  * provide feedback on https://github.com/quarkusio/quarkus/issues[GitHub];

Reviewer 5: ```suggestion
 * Provide feedback on https://github.com/quarkusio/quarkus/issues[GitHub].
```

>  * craft some code and https://github.com/quarkusio/quarkus/pulls[push a PR];

Reviewer 5: ```suggestion
 * Craft some code and https://github.com/quarkusio/quarkus/pulls[push a PR].
```

>  * discuss with us on https://quarkusio.zulipchat.com/[Zulip] and on the https://groups.google.com/d/forum/quarkus-dev[mailing list];

Reviewer 5: ```suggestion
 * Discuss with us on https://quarkusio.zulipchat.com/[Zulip] and on the https://groups.google.com/d/forum/quarkus-dev[mailing list].
```

>  * ask your questions on https://stackoverflow.com/questions/tagged/quarkus[Stack Overflow].

Reviewer 5: ```suggestion
 * Ask your questions on https://stackoverflow.com/questions/tagged/quarkus[Stack Overflow].
```
